src/raftstore/store/peer.rs

```diff
@@ -1875,6 +1800,123 @@ impl Peer {
     }
 }
 
+/// Doctrine decides how we handle a request.
+enum RequestDoctrine {
+    // Handle the read request directly without dispatch.
+    ReadLocal,
+    // Handle the read request via raft's SafeReadIndex mechanism.
+    ReadIndex,
+    ProposeNormal,
+    ProposeTransferLeader,
+    ProposeConfChange,
+}
+
+/// A jury helps `RequestJudge` make a `RequestDoctrine`.
+trait RequestJury {
+    /// Has the jury applied the current term?
+    fn has_applied_current_term(&self) -> bool;
+    /// Inspects the jury's lease.
+    fn inspect_lease(&self) -> LeaseState;
+}
+
+/// Judge makes `RequestDoctrine` for requests.
+struct RequestJudge<'a, T: 'a + RequestJury> {
+    jury: &'a T,
+}
+
+impl<'a, T: RequestJury> RequestJudge<'a, T> {
+    /// Create a `RequestJudge`.
+    fn with_jury(jury: &'a T) -> RequestJudge<T> {
```

Comment:
You can use `impl RequestJury` here now I think. :)

Reply:
Well, I think `&'a T` is fine, I like the idea of [the RFC]( https://github.com/phaazon/rfcs/blob/undo-universal-impl-trait/text/0000-undo-universal-impl-trait.md#why-is-impl-trait-in-argument-position-wrong) .

Reply:
Ok :)